clients/src/main/java/org/apache/kafka/clients/MetadataCache.java

```diff
@@ -52,7 +52,7 @@ public class MetadataCache {
     private final Map<TopicPartition, PartitionMetadata> metadataByPartition;
     private final Map<String, Uuid> topicIds;
     private final Map<Uuid, String> topicNames;
-    private Cluster clusterInstance;
+    private InternalCluster clusterInstance;
```

Comment:
The javadoc for `MetadataCache` describes it as mutable, but as far as I can tell, we do not actually modify any of the collections. We always build new instances instead of updating an existing one. That makes me wonder if we can change the javadoc and use `MetadataCache` as the immutable snapshot of metadata state. Then we could drop `InternalCluster` in favor of `MetadataCache`. Would that work?

Reply:
@hachikuji I had thought about `MetadataCache`. It has 1 accessor i.e. `partitionMetadata()` that is returning mutable `PartitionMetadata`, and is not making defensive copies. Rest all accessors are returning immutable objects, or making defensive copies. 
Is it ok for `partitionMetadata()` to make defensive copies? That would lead to memory going up.

Reply:
Discussed offline. It does not look like `PartitionMetadata` should be treated as mutable. It comes directly from the Metadata response and I can't think of a reason the client would update any of the replica sets directly. We should confirm though.

Reply:
> We should confirm though.

I don't see it being used mutably in code. I see historically, it was made mutable to support deletion/updates within cache, but the deletion/update code has since been removed. As far i can see, read-only semantic. So i have treated `MetadataCache` as immutable cache, made its internal data structures unmodifiable and updated the javadoc.

All clients test pass locally, hopefully Jenkins signal is green too 🤞